### PR TITLE
[RFC] Minor cleanups

### PR DESCRIPTION
**Pull request by tonyo**
Here are a few minor cleanups. Issues were discovered by static analysis (mainly by PVS-Studio).

Summary of changes:
- Remove expressions which always evaluate to TRUE in their context in `regexp_nfa.c` and `search.c` (`len == 0` and `!define_matched`)
- Merge consecutive identical 'if' statements in `farsi.c` and `fileio.c`
- Remove redundant parentheses in `tag.c`
- Fix an out-of-memory cleanup in `memline.c` (`i` was used for both outer and inner loops)
- Remove unnecessary `free()` from `menu.c` (`from_noamp` is always NULL there)
- Call `f_browse()` from  `f_browsedir` (since they are identical)

### CODE REVIEW

File: src/nvim/memline.c [hunk 1]
```diff
@@ -1360,9 +1360,10 @@ recover_names (
     /* check for out-of-memory */
     for (i = 0; i < num_names; ++i) {
       if (names[i] == NULL) {
-        for (i = 0; i < num_names; ++i)
-          free(names[i]);
+        for (j = 0; j < num_names; ++j)
+          free(names[j]);
         num_names = 0;
+        break;
```
Comment (justinmk) [inline src/nvim/memline.c]: Is this actually a bug, or just to make the intent clear?

File: src/nvim/memline.c [hunk 2]
```diff
@@ -1277,7 +1277,7 @@ recover_names (
   int num_files;
   int file_count = 0;
   char_u      **files;
-  int i;
+  int i, j;
```
Comment (justinmk) [inline src/nvim/memline.c]: Would you mind moving this to local block scope for each `for(int i = 0; ...)` block in this function?

File: src/nvim/eval.c [hunk 3]
```diff
@@ -7287,8 +7287,7 @@ static void f_browse(typval_T *argvars, typval_T *rettv)
  */
 static void f_browsedir(typval_T *argvars, typval_T *rettv)
 {
-  rettv->vval.v_string = NULL;
-  rettv->v_type = VAR_STRING;
+  f_browse(argvars, rettv);
```
Comment (justinmk) [inline src/nvim/eval.c]: This is weird. Turns out we removed `FEAT_BROWSE`. I think we should actually restore that. https://github.com/neovim/neovim/issues/2160

**Comment by justinmk** [comment]
Nice cleanup!

File: src/nvim/memline.c (hunk 1 above)
Comment (rev112) [inline src/nvim/memline.c]: I believe the original code does its job, but I needed some time to figure out why it had been written this way. So yes, to make the intent clear.

File: src/nvim/eval.c (hunk 3 above)
Comment (rev112) [inline src/nvim/eval.c]: I don't know anything about `FEAT_BROWSE` yet, sorry. So this means that I leave this place as is for now?

Comment (justinmk) [inline src/nvim/eval.c]: It means we need to figure what it means :) you can leave this change as is.

**Comment by rev112** [comment]
@justinmk, @Pyrohh, thanks for your comments, I've updated the PR.

**Comment by justinmk** [comment]
merged, thanks @rev112

**Comment by Pyrohh** [comment]
Thanks @rev112. By the way, did you pay for PVS studio? I heard the creator is quite nice and generally will give out licenses for open source projects if you email him/her, but I don't know if that's true.

**Comment by rev112** [comment]
I used the free trial version of PVS-Studio Standalone which has a limited number of "clicks" (basically, operations in the GUI window of the tool, more info here: http://www.viva64.com/en/d/0012/).
I heard good things about the creator/company as well, but most likely the license will be quite limited in time (probably around 1 month, as written here: http://www.viva64.com/en/b/0092/). I can write them and ask if there's interest.

**Comment by Pyrohh** [comment]
> I can write them and ask if there's interest.

It's fine, but thanks. That's all I needed to know.

**Comment by justinmk** [comment]
@chrisbra FYI this PR addressed some PVS studio items (which are present in Vim).

**Comment by chrisbra** [comment]
On Do, 19 Mär 2015, Justin M. Keyes wrote:

> @chrisbra FYI this PR addressed some PVS studio items (which are present in Vim).

Thanks.

**Pull request closed without merging**